### PR TITLE
add option of passing identity string directly

### PR DESCRIPTION
**Pull request by cerkiewny**
does this work for issue #50?

### CODE REVIEW

File: demo/demo_args.c [hunk 1]
```diff
@@ -1125,6 +1128,19 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
         case 305:
             parsed_args->alternative_logger = true;
             break;
+        case 306: {
+            if(!optarg) {
+                demo_log(ERROR, "Missing identity string");
+                goto finish;
+            }
```
Comment (kFYatek) [inline demo/demo_args.c:1135]: This would be redundant if you used `required_argument`

**Review by kFYatek — CHANGES_REQUESTED** [review]
Thanks for this PR, it does indeed fix that issue. There are however a couple of things that could be improved upon.

File: demo/demo_args.c (hunk 1 above)
Comment (cerkiewny) [reply demo/demo_args.c:1135]: Oh, I thought required argument does something else :D (as in you need to provide it to the executable in order for it to be happy, it is more like "requires value", right?)

File: demo/demo_args.c [hunk 2]
```diff
@@ -1125,6 +1128,19 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
...
+            const size_t identity_lenght = strlen(optarg);
```
[the remark this answers is not in the document]
Comment (cerkiewny) [reply demo/demo_args.c:1136]: done

File: demo/demo_args.c [hunk 3]
```diff
@@ -1125,6 +1128,19 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
...
+            if(parsed_args->connection_args.public_cert_or_psk_identity != NULL || identity_lenght == 0) {
+                demo_log(ERROR, "Invalid identity either identity was set twice or empty parameter was passed");
+                goto finish;
+            }
+            clone_buffer(&parsed_args->connection_args.public_cert_or_psk_identity, &parsed_args->connection_args.public_cert_or_psk_identity_size, optarg, identity_lenght);
+            break;
+        }
```
[the remark this answers is not in the document]
Comment (cerkiewny) [reply demo/demo_args.c:1143]: done

File: demo/demo_args.c [hunk 4]
```diff
@@ -611,6 +613,7 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
         { "dm-persistence-file",           required_argument, 0, 289 },
 #endif // defined(AVS_COMMONS_WITH_AVS_PERSISTENCE) && defined(AVS_COMMONS_STREAM_WITH_FILE)
         { "alternative-logger",            no_argument,       0, 305 },
+        { "identitystring",                optional_argument, 0, 306 },
```
[the remark this answers is not in the document]
Comment (cerkiewny) [reply demo/demo_args.c:616]: 2 top ones done, maybe let's open another issue for the remaining params ?

[the remark this answers is not in the document]
Comment (kFYatek) [reply demo/demo_args.c]: Doing `fw_security_info` in a separate issue/PR would make sense, but identity and PSK are usually passed together, so I think that adding support for at least `parsed_args->connection_args.private_cert_or_psk_key` (currently accessible through `-k`/`--key`) in the same PR would be preferrable.

[the remark this answers is not in the document]
Comment (cerkiewny) [reply demo/demo_args.c]: sure, makes sense

File: demo/demo_args.c [hunk 5]
```diff
@@ -1125,6 +1131,36 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
         case 305:
             parsed_args->alternative_logger = true;
             break;
+        case 306: {
+            const size_t identity_length = strlen(optarg);
+            if (parsed_args->connection_args.public_cert_or_psk_identity != NULL
+                    || identity_length == 0) {
+                demo_log(ERROR, "Invalid identity either identity was set "
```
Comment (kFYatek) [inline demo/demo_args.c:1138]: Nitpick: missing punctuation

File: demo/demo_args.c [hunk 6]
```diff
@@ -1125,6 +1131,36 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
...
+            if (parsed_args->connection_args.private_cert_or_psk_key != NULL
+                    || key_length == 0) {
+                demo_log(ERROR, "Invalid key, either key was set "
+                                "twice or empty parameter was passed");
+                goto finish;
+            }
+            clone_buffer(
+                    &parsed_args->connection_args.private_cert_or_psk_key,
+                    &parsed_args->connection_args
+                             .private_cert_or_psk_key_size,
```
Comment (kFYatek) [inline demo/demo_args.c:1160]: Nitpick: this can fit in 80 columns without line wrap

File: demo/demo_args.c [hunk 7]
```diff
@@ -325,6 +325,10 @@ static void print_help(const struct option *options) {
         { 305, NULL, NULL,
           "Enable alternative logger as a showcase of extended logger "
           "feature." },
+        { 306, NULL, NULL,
+          "Provide identity from ASCII string (see -i parameter for more details)" },
```
Comment (kFYatek) [inline demo/demo_args.c:329]: Nitpick: this exceeds 80 columns

File: demo/demo_args.c [hunk 8]
```diff
@@ -1125,6 +1131,36 @@ int demo_parse_argv(cmdline_args_t *parsed_args, int argc, char *argv[]) {
...
+                    || identity_length == 0) {
+                demo_log(ERROR, "Invalid identity either identity was set "
+                                "twice or empty parameter was passed");
+                goto finish;
+            }
+            clone_buffer(
+                    &parsed_args->connection_args.public_cert_or_psk_identity,
+                    &parsed_args->connection_args
+                             .public_cert_or_psk_identity_size,
+                    optarg, identity_length);
```
Comment (kFYatek) [inline demo/demo_args.c:1146]: Here and in the usage below: you don't check the return value of `clone_buffer()`, so in the (unlikely) case of an out-of-memory condition, the user won't even get an error message.

**Review by kFYatek — CHANGES_REQUESTED** [review]
Thanks, this looks pretty good already! There's just this unchecked out-of-memory case that I missed in the last review.

File: demo/demo_args.c (hunk 6 above)
Comment (cerkiewny) [reply demo/demo_args.c:1160]: Is there any chance of getting clang-format file ? :D

Comment (kFYatek) [reply demo/demo_args.c:1160]: Here's the clang-format file: https://gist.github.com/kFYatek/8b716be6e6fff080a21c97101f7a97d6
However it targets a custom fork of clang-format 6.0 that you can get from https://github.com/kFYatek/clang or https://github.com/dextero/clang/tree/format/dextero... for those changes you're making it's likely that upstream clang-format will work as well, but you'll probably need to remove some of the config options.

**Comment by cerkiewny** [comment]
@kFYatek I think you will have to re-format the patch, I am running into some weird problems, either the clang-format-6.0 doesn't understand some things, or clang-format-10.0 does too much and aligns ifs differently...

**Comment by kFYatek** [comment]
Rebased, reformatted and merged in https://github.com/AVSystem/Anjay/commit/d5debe4fe3b4b42650d796d7c66bf8d86a7f0ffe. Thanks!

**Pull request closed without merging**